mediaapi/routing/download.go

```diff
@@ -767,19 +826,94 @@ func (r *downloadRequest) fetchRemoteFile(
+	resp, err := r.fedClient.DownloadMedia(ctx, r.origin, r.MediaMetadata.Origin, string(r.MediaMetadata.MediaID))
 	if err != nil || (resp != nil && resp.StatusCode != http.StatusOK) {
-		if resp != nil && resp.StatusCode == http.StatusNotFound {
-			return "", false, fmt.Errorf("File with media ID %q does not exist on %s", r.MediaMetadata.MediaID, r.MediaMetadata.Origin)
+		isMultiPart = false
+		// try again on the unauthed endpoint
+		// create request for remote file
+		resp, err = client.CreateMediaDownloadRequest(ctx, r.MediaMetadata.Origin, string(r.MediaMetadata.MediaID))
+		if err != nil || (resp != nil && resp.StatusCode != http.StatusOK) {
+			if resp != nil && resp.StatusCode == http.StatusNotFound {
+				return "", false, fmt.Errorf("File with media ID %q does not exist on %s", r.MediaMetadata.MediaID, r.MediaMetadata.Origin)
+			}
+			return "", false, fmt.Errorf("file with media ID %q could not be downloaded from %s", r.MediaMetadata.MediaID, r.MediaMetadata.Origin)
 		}
-		return "", false, fmt.Errorf("file with media ID %q could not be downloaded from %s", r.MediaMetadata.MediaID, r.MediaMetadata.Origin)
 	}
 	defer resp.Body.Close() // nolint: errcheck
 
-	// The reader returned here will be limited either by the Content-Length
-	// and/or the configured maximum media size.
-	contentLength, reader, parseErr := r.GetContentLengthAndReader(resp.Header.Get("Content-Length"), &resp.Body, maxFileSizeBytes)
+	// If this wasn't a multipart response, set the Content-Type now. Will be overwritten
+	// by the multipart Content-Type below.
+	r.MediaMetadata.ContentType = types.ContentType(resp.Header.Get("Content-Type"))
+
+	var contentLength int64
+	var reader io.Reader
+	var parseErr error
+	if isMultiPart {
+		r.Logger.Debug("Downloaded file using authenticated endpoint")
+		var params map[string]string
+		_, params, err = mime.ParseMediaType(resp.Header.Get("Content-Type"))
+		if err != nil {
+			return "", false, err
+		}
+		if params["boundary"] == "" {
+			return "", false, fmt.Errorf("no boundary header found on media %s from %s", r.MediaMetadata.MediaID, r.MediaMetadata.Origin)
+		}
+		mr := multipart.NewReader(resp.Body, params["boundary"])
+
+		// Get the first, JSON, part
+		p, multipartErr := mr.NextPart()
+		if multipartErr != nil {
+			return "", false, multipartErr
+		}
+
+		if p.Header.Get("Content-Type") != "application/json" {
+			return "", false, fmt.Errorf("first part of the response must be application/json")
+		}
+		// Try to parse media meta information
+		meta := mediaMeta{}
+		if err = json.NewDecoder(p).Decode(&meta); err != nil {
+			return "", false, err
+		}
+		defer p.Close() // nolint: errcheck
+
+		// Get the actual media content
+		p, multipartErr = mr.NextPart()
+		if multipartErr != nil {
+			return "", false, multipartErr
+		}
+
+		redirect := p.Header.Get("Location")
+		if redirect != "" {
+			if !strings.HasPrefix(redirect, "https://") {
+				return "", false, fmt.Errorf("redirect URL must be HTTPS")
```

Comment:
Federation has to be all https, but I guess redirects could allow http too (should maybe have that in the spec 🤔). The main thing I wanted to prevent is someone using non-http things like the special `matrix://` protocol

Reply:
`Location` header is now removed, due to not having any mechanism to allow/block access to certain resources.